### PR TITLE
[ConstraintSystem] Match trailing closures to implicit `.callAsFunction` when necessary

### PR DESCRIPTION
**Pull request by xedin**
In situations like `T(...) { ... }` where `T` is a callable type,
it's not immediately clear whether trailing closures are associated
with `.init` of `T` or implicit `.callAsFunction`. So if constructor
didn't match trailing closures, let's attempt to split them off and
form an implicit `.callAsFunction` call with them as a fallback.

Resolves: rdar://83717297

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -10634,15 +10663,106 @@ ConstraintSystem::simplifyApplicableFnConstraint(
...
+                                  callableType, func1->getExtInfo());
+
+        auto matchCallResult = ::matchCallArguments(
+            *this, func2, newArgumentList, func1->getParams(),
+            func2->getParams(), subKind, argumentsLoc, trailingClosureMatching);
+
+        if (matchCallResult != SolutionKind::Solved)
+          return SolutionKind::Error;
+
+        // Note that `createImplicit` cannot be used here because it
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:10739]: I see `ArgumentList::createParsed` do set implicit as `false`, so I wonder If it make sense in this case `setImplicit()` in `implicitCallArgumentList`?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -1341,7 +1342,6 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
   ParameterListInfo paramInfo(params, callee, appliedSelf);
 
   // Dig out the argument information.
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/CSSimplify.cpp:1344]: Ah yeah, I removed the code but not comment of course :)

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp:10739]: Yeah, I think it's worth it, thanks!

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:10739]: Heh, there is no `setImplicit()` in argument list which is intentional, so I'm just going to add trailing index to `createImplicit` looks like there is no way around that.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:10739]: I see `ArgumentList::createParsed` just call `ArgumentList::create`(which is also public) that takes implicit as parameter, so maybe call `create` directly? :)

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:10739]: Nah, `createImplicit` makes sense here, I just didn't want to add a parameter but that's fine too :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:10739]: Ah awesome, thanks!

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:10739]: Fixed!

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Fixed!

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:10739]: I think we forgot the comment :)

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Good catch, fixed!

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/CSSimplify.cpp [hunk 3]
```diff
@@ -10634,15 +10663,103 @@ ConstraintSystem::simplifyApplicableFnConstraint(
...
+        SmallVector<Argument, 4> newArguments;
+        SmallVector<Argument, 4> trailingClosures;
+
+        for (unsigned i = 0, n = argumentList->size(); i != n; ++i) {
+          if (argumentList->isTrailingClosureIndex(i)) {
+            trailingClosures.push_back(argumentList->get(i));
+          } else {
+            newArguments.push_back(argumentList->get(i));
+          }
+        }
```
Comment (hamishknight) [inline lib/Sema/CSSimplify.cpp:10708]: FWIW `ArgumentList` exposes iterator ranges for this e.g:

```cxx
SmallVector<Argument, 4> newArguments(argumentList->getNonTrailingArgs());
SmallVector<Argument, 4> trailingClosures(argumentList->getTrailingClosures());
```

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:10708]: For some reason I thought that SmallVector would only accept begin+end and decided to just iterate instead…

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Adjusted, thank you!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

File: lib/Sema/CSSimplify.cpp [hunk 4]
```diff
@@ -10647,8 +10647,108 @@ ConstraintSystem::simplifyApplicableFnConstraint(
         subKind, argumentsLoc, trailingClosureMatching);
 
     switch (matchCallResult) {
-    case SolutionKind::Error:
+    case SolutionKind::Error: {
+      if (shouldAttemptFixes())
+        return SolutionKind::Error;
```
Comment (hborla) [inline lib/Sema/CSSimplify.cpp]: Hmm, does this mean we won't attempt to re-match the trailing closure to `callAsFunction` in diagnostic mode? Will that lead to any misleading diagnostics if there are other errors in the closure or something like that?

**Review by hborla — APPROVED** [review]
LGTM! I had one question about matching trailing closures to `callAsFunction` in diagnostic mode, but it looks like you already fixed it in a later commit.

File: lib/Sema/CSSimplify.cpp (hunk 4 above)
Comment (hborla) [reply lib/Sema/CSSimplify.cpp]: Looks like this is changed in https://github.com/apple/swift/pull/41189/commits/1c258ca05f37b7081765df742262887291c41fc6

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Yeah, sorry I was testing changes and forgot to remove it initially and indeed injection should be attempted in both modes.

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

**Comment by xedin** [comment]
@swift-ci please build toolchain macOS

**Comment by xedin** [comment]
Ok, I give up, going to merge and wait for the main branch toolchain to be built.

**Pull request merged**